File: Merge-2-sorted-array.cpp
// Time Complexity : O(m+n)
// Space Complexity : O(m)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : yes

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        if(nums1.size()==0){
            return;
        }
        int i=m+n-1, p1 = m-1, p2 = n-1;
        while(p1>=0 && p2>=0){
            if(nums1[p1]>nums2[p2]){
                nums1[i] = nums1[p1];
                p1--;
            }
            else{
                nums1[i] = nums2[p2];
                p2--;
            }
            i--;
        }
        while(p2 >= 0){
            nums1[i] = nums2[p2];
            p2--;
            i--;
        }
    }
};

File: merge2SortedArray.java
/*
Run time complexity - O(N+M)
Space complexity - O(1)
*/
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        if(n==0)
            return;
        int temp=m+n-1;
        for(int i=m-1;i>=0;i--)
        {
            nums1[temp]=nums1[i];
            nums1[i]=0;
            temp--;
        }
        int a=n;
        int b=0;
        int pos=0;
        while(b<n && a<(m+n))
        {
            if(nums1[a]<=nums2[b])
            {
                nums1[pos]=nums1[a];
                nums1[a]=0;
                a++;
            }
            else if(nums2[b]<nums1[a])
            {
                nums1[pos]=nums2[b];
                b++;
            }
            pos++;
        }
        while(b<n)
        {
            nums1[pos]=nums2[b];
            b++;
            pos++;
        }
        //nums1 will already be in it's own position
    }
}
